blob/azureblob/azureblob.go

```diff
@@ -565,16 +585,33 @@ type writer struct {
 // NewTypedWriter implements driver.NewTypedWriter.
 func (b *bucket) NewTypedWriter(ctx context.Context, key string, contentType string, opts *driver.WriterOptions) (driver.Writer, error) {
 	blockBlobURL := b.blockBlobURL(key)
-	if opts.Metadata == nil {
-		opts.Metadata = map[string]string{}
-	}
 	if opts.BufferSize == 0 {
 		opts.BufferSize = defaultUploadBlockSize
 	}
+
+	md := make(map[string]string, len(opts.Metadata))
+	for k, v := range opts.Metadata {
+		// See the package comments for more details on escaping of metadata
+		// keys & values.
+		e := escape.Escape(k, func(runes []rune, i int) bool {
+			c := runes[i]
```

Comment:
Is the rule that the keys must be valid _identifiers_, or that they must only contain valid identifier _characters_?

The difference would be in keys that start with a digit. Are they legal?

Reply:
Good catch. No, they are not. I added a test to `escape.WeirdStrings` and check for that here now.